backoff.c

```diff
@@ -0,0 +1,87 @@
+#include "common.h"
+
+#include <ext/standard/php_rand.h>
+#include <ext/standard/php_mt_rand.h>
+
+#include "backoff.h"
+
+zend_long random_range(zend_long min, zend_long max) {
+    if (max <= min) {
+        max = min + 1;
+    }
+
+  return php_mt_rand_range(min, max);
+}
+
+zend_long default_backoff(struct Backoff *self, int retry_index) {
+    zend_long backoff = retry_index ? self->base : random_range(0, self->base);
+    return MIN(self->cap, backoff);
+}
+
+zend_long constant_backoff(struct Backoff *self, int retry_index) {
+    zend_long backoff = self->base;
+    return MIN(self->cap, backoff);
+}
+
+zend_long uniform_backoff(struct Backoff *self, int retry_index) {
+  zend_long backoff = random_range(0, self->base);
+    return MIN(self->cap, backoff);
+}
+
+zend_long exponential_backoff(struct Backoff *self, int retry_index) {
+    zend_long pow = MIN(retry_index, 10);
+    zend_long backoff = self->base * (1 << pow);
+    return MIN(self->cap, backoff);
+}
+
+zend_long full_jitter_backoff(struct Backoff *self, int retry_index) {
+    zend_long pow = MIN(retry_index, 10);
+    zend_long backoff = self->base * (1 << pow);
+    zend_long cap = MIN(self->cap, backoff);
+    return random_range(0, self->cap);
+}
+
+zend_long equal_jitter_backoff(struct Backoff *self, int retry_index) {
+    zend_long pow = MIN(retry_index, 10);
+    zend_long backoff = self->base * (1 << pow);
+    zend_long temp = MIN(self->cap, backoff) >> 1;
+    return temp + random_range(0, temp);
```

Comment:
YMMV but I think it's easier to follow what's going on if we just divide rather than shift.

```suggestion
    zend_long temp = MIN(self->cap, backoff);
    return temp / 2 + random_range(0, temp / 2);
```

Reply:
Thanks. Fixed.